proposals/0436-recordwildcards-total.rst

```diff
@@ -0,0 +1,102 @@
+.. ticket-url:: Leave blank. This will eventually be filled with the
+                ticket URL which will track the progress of the
+                implementation of the feature.
+.. implemented:: Leave blank. This will be filled in with the first GHC version which
+                 implements the described feature.
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/436>`_.
+.. contents::
+
+``-XRecordWildCards`` does not give any warning or indication if you forget to use a field. This proposal adds syntax (optionally enabled by another extension) to give a warning if a field is unused.
+
+
+Motivation
+----------
+
+Some functions require all the fields of a record to be used, or at least considered, e.g. ``hash`` or ``toJSON``, or even something more security-facing. Currently, if one uses ``RecordWildCards`` and they add a field to the record later on, nothing will tell the developer to take a look at these functions.
+
+::
+
+ data Foo = Foo { a :: Int, b :: Bool }
+ 
+ instance ToJSON Foo where
+   toJSON Foo{..} =
+     object
+       [ "a" .= a
+       , "b" .= b
+       ]
+       
+ instance Hashable Foo where
+   hash Foo{..} = hash (a, b)
+ 
+ -- adding a 'c' field does not raise any indication of needing to look at `toJSON` or `hash`
+
+Existing workarounds are described in `this blog post <https://cs-syd.eu/posts/2021-09-10-undefined-trick>`_, all of which leave something to be desired:
+
+* Positional arguments break when fields are reordered, which matters in functions like `toJSON`
+* Using `undefined` is a bit of a hack, plus it fails `-Wunused-pattern-binds`. It also fails to explicitly say which field is unused (which isn't strictly necessary, but it's nice to have)
+
+The `safe-wild-cards <https://hackage.haskell.org/package/safe-wild-cards>`_ package provides a Template Haskell solution to this, but it would be nice to support this natively.
+
+One could even think of this as the dual of the `Or pattern proposal <https://github.com/ghc-proposals/ghc-proposals/pull/43>`_, to warn on unused terms in product types as opposed to unused terms in sum types.
+
+Proposed Change Specification
+-----------------------------
+
+When ``-XRecordWildCards`` is enabled, in addition to supporting ``Foo{..}`` in a pattern match, also support ``Foo{..!}``, which desugars to ``Foo{a = a, b = b}``.
+
+Examples
+--------
+
+Now, the example in the Motivation section could be written as
+
+::
+
+ instance ToJSON Foo where
+   toJSON Foo{..!} =
+     object
+       [ "a" .= a
+       , "b" .= b
+       ]
+       
+ instance Hashable Foo where
+   hash Foo{..!} = hash (a, b)
+   
+Notice the lack of any change, except for swapping ``Foo{..}`` for ``Foo{..!}``.
+
+
+Effect and Interactions
+-----------------------
+The desugared syntax would raise an ``-Wunused-matches`` warning if a new field is added but not used.
```

Comment:
Why can `-Wunused-matches` not simply be changed to trigger for any unused fields (as this proposal assumes a new syntax is needed to do)? Because it would trigger on too much of Hackage? I'd like to see this addressed. I don't think it is so bad to expand the scope of a warning, it doesn't even break the build, unless they have opted in to that with `-Werror`. If you think the existing warning cannot be changed, why not add a new warning?

I would in fact find a warning based approach advantageous, since GHC/HLS could suggest a fix for it, which would be to add the NamedFieldPuns extension and expand the `..` to the list of fields that are currently used. With HLS, it could be a one-click deal. I don't know if HLS policy allows suggesting new extensions for syntax that is *not* currently used, that is. Though I think it would make sense.

Since NamedFieldPuns is part of GHC2021, and RecordWildCards is not, I think it would be good to nudge people in that direction.

Reply:
Because in most* cases, you don't intend to use all the fields in RecordWildCards, you just don't want to list out all the fields. For example, if you have a Reader monad environment where the environment has many fields (e.g. ours has 10), you might use most but not all in one function.

This proposal is specifically for the use-case of functions like `hash` or `toJSON`, where the developer would like to be strict about the fact that the function _must_ account for all the fields. But in other functions, you're just using RecordWildCards to bring all the fields into scope, whether you use it or not